### PR TITLE
MIG 9.0 account_invoice_tax_required

### PR DESCRIPTION
**Pull request by angelmoya**

### CODE REVIEW

File: account_invoice_tax_required/README.rst [hunk 1]
```diff
@@ -1,19 +1,32 @@
...
 
 This module adds functional a check on invoice
 to force user to set tax on invoice line.
 
+Usage
+=====
+
+To use this module, you need to:
+
+#. Go to ...
```
Comment (pedrobaeza) [inline account_invoice_tax_required/README.rst]:
```
#. Go to Accounting (or Invoicing) > Sales > Customer Invoices.
#. Create a new invoice.
#. Insert a new invoice line without taxes.
#. Press on "Validate".
#. An error will raise preventing to do it.
```

File: account_invoice_tax_required/README.rst [hunk 2]
```diff
@@ -22,17 +35,19 @@ Contributors
 ------------
 
 * Vincent Renaville <vincent.renaville@camptocamp.com>
+* Angel Moya <odoo@tecnativa.com>
```
Comment (pedrobaeza) [inline account_invoice_tax_required/README.rst:43]: Don't you put `Ángel`?

File: account_invoice_tax_required/tests/test_account_invoice_tax_required.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
+# Copyright <YEAR(S)> <AUTHOR(S)>
```
Comment (pedrobaeza) [inline account_invoice_tax_required/tests/test_account_invoice_tax_required.py]: Put here the copyright

File: account_invoice_tax_required/tests/test_account_invoice_tax_required.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp.tests.common import TransactionCase
+from openerp import exceptions
+
+
+class TestAccountInvoiceTaxRequired(TransactionCase):
+
+    def setUp(self, *args, **kwargs):
```
Comment (pedrobaeza) [inline account_invoice_tax_required/tests/test_account_invoice_tax_required.py]: This is only `def setUp(self)`

File: account_invoice_tax_required/tests/test_account_invoice_tax_required.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        super(TestAccountInvoiceTaxRequired, self).setUp(*args, **kwargs)
+        self.main_company = self.env.ref('base.main_company')
+        res_users_account_manager = self.env.ref(
```
Comment (pedrobaeza) [inline account_invoice_tax_required/tests/test_account_invoice_tax_required.py]: There's no need to create a user for this. This test can be executed with admin.

File: account_invoice_tax_required/tests/test_account_invoice_tax_required.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+                name="Adviser",
+                company_id=self.main_company.id,
+                login="fm",
+                email="accountmanager@yourcompany.com",
+                groups_id=[(6, 0,
+                            [res_users_account_manager.id,
+                             partner_manager.id])]
+            ))
+        self.account_invoice_obj = self.env['account.invoice']
+        self.payment_term = self.env.ref(
```
Comment (pedrobaeza) [inline account_invoice_tax_required/tests/test_account_invoice_tax_required.py]: Don't look for any payment term. It's not needed for the invoice completion.

File: account_invoice_tax_required/tests/test_account_invoice_tax_required.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            'account.account_payment_term_advance')
+        self.journalrec = self.env['account.journal'].search(
```
Comment (pedrobaeza) [inline account_invoice_tax_required/tests/test_account_invoice_tax_required.py]: In this case, I prefer to create a new journal for the test.

File: account_invoice_tax_required/tests/test_account_invoice_tax_required.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                code="cust_acc",
+                name="customer account",
+                user_type_id=account_user_type.id,
+                reconcile=True,
+            ))
+
+        invoice_line_data = [
+            (0, 0,
+                {
+                    'product_id': self.env.ref('product.product_product_5').id,
```
Comment (pedrobaeza) [inline account_invoice_tax_required/tests/test_account_invoice_tax_required.py]: Create a test product. Don't use demo data.

File: account_invoice_tax_required/tests/test_account_invoice_tax_required.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+                          '=',
+                          self.env.ref('account.data_account_type_revenue').id)
+                         ], limit=1).id,
+                    'name': 'product test 5',
+                    'price_unit': 100.00,
+                }
+             )
+        ]
+
+        self.invoice = self.account_invoice_obj.sudo(
```
Comment (pedrobaeza) [inline account_invoice_tax_required/tests/test_account_invoice_tax_required.py]: sudo is not required, as you are already executing this as admin.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: account_invoice_tax_required/README.rst (hunk 2 above)
Comment (angelmoya) [reply account_invoice_tax_required/README.rst:43]: I usually put unaccent, on my DNI (Spanish Identity Card) is writen unaccent, and also in all oficial documents. I know that it is wrong write on spanish, but unaccent is more international ;)

Comment (pedrobaeza) [reply account_invoice_tax_required/README.rst:43]: OK, it was just curiosity.

File: account_invoice_tax_required/tests/test_account_invoice_tax_required.py (hunk 9 above)
Comment (angelmoya) [reply account_invoice_tax_required/tests/test_account_invoice_tax_required.py]: I Fixed

**Comment by angelmoya** [comment]
@pedrobaeza all comments fixed

**Comment by pedrobaeza** [comment]
Some pylint errors:

```
account_invoice_tax_required/tests/test_account_invoice_tax_required.py:27:17: E126 continuation line over-indented for hanging indent
account_invoice_tax_required/tests/test_account_invoice_tax_required.py:31:13: E121 continuation line under-indented for hanging indent
account_invoice_tax_required/tests/test_account_invoice_tax_required.py:34:21: E126 continuation line over-indented for hanging indent
account_invoice_tax_required/tests/test_account_invoice_tax_required.py:62:17: E126 continuation line over-indented for hanging indent
account_invoice_tax_required/tests/test_account_invoice_tax_required.py:68:13: E121 continuation line under-indented for hanging indent
```

You can avoid this installing this pre-commit hook on GIT: https://github.com/OCA/maintainer-quality-tools/tree/master/git

**Comment by pedrobaeza** [comment]
cc @Tecnativa

File: account_invoice_tax_required/models/account_invoice.py [hunk 10]
```diff
@@ -28,14 +14,15 @@ def test_invoice_line_tax(self):
         errors = []
         error_template = _("Invoice has a line with product %s with no taxes")
         for invoice in self:
```
Comment (pedrobaeza) [inline account_invoice_tax_required/models/account_invoice.py]: You can avoid one loop with:

```
for invoice_line in self.mapped('invoice_line'):
    ...
```

File: account_invoice_tax_required/models/account_invoice.py [hunk 11]
```diff
@@ -28,14 +14,15 @@ def test_invoice_line_tax(self):
         errors = []
```
Comment (pedrobaeza) [inline account_invoice_tax_required/models/account_invoice.py:15]: We should ignore this test on 2 cases:
- With demo data. You can see the problem on current runbot build:
  <details>
  No Taxes Defined!
  Invoice has a line with product Dell Inspiron Laptop without OS with no taxes,Invoice has a line with product Linutop with no taxes
  None
  Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/tools/yaml_import.py", line 884, in process
    self._process_node(node)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/tools/yaml_import.py", line 916, in _process_node
    self.process_workflow({node: []})
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/tools/yaml_import.py", line 655, in process_workflow
    openerp.workflow.trg_validate(uid, workflow.model, id, workflow.action, self.cr)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/workflow/__init__.py", line 67, in trg_validate
    return WorkflowService.new(cr, uid, res_type, res_id).validate(signal)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/workflow/service.py", line 73, in validate
    res2 = wi.validate(signal)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/workflow/instance.py", line 62, in validate
    wi.process(signal=signal, force_running=force_running, stack=stack)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/workflow/workitem.py", line 102, in process
    ok = self._split_test(activity['split_mode'], signal, stack)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/workflow/workitem.py", line 230, in _split_test
    self._join_test(t[0], t[1], stack)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/workflow/workitem.py", line 239, in _join_test
    WorkflowItem.create(self.session, self.record, activity, inst_id, stack=stack)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/workflow/workitem.py", line 77, in create
    workflow_item.process(stack=stack)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/workflow/workitem.py", line 98, in process
    if not self._execute(activity, stack):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/workflow/workitem.py", line 144, in _execute
    returned_action = self.wkf_expr_execute(activity)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/workflow/workitem.py", line 295, in wkf_expr_execute
    return self.wkf_expr_eval_expr(activity['action'])
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/workflow/workitem.py", line 273, in wkf_expr_eval_expr
    result = eval(line, env, nocopy=True)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/tools/safe_eval.py", line 298, in safe_eval
    return eval(c, globals_dict, locals_dict)
  File "", line 1, in <module>
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/api.py", line 248, in wrapper
    return new_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/addons/account_invoice_tax_required/models/account_invoice.py", line 31, in invoice_validate
    self.test_invoice_line_tax()
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/api.py", line 248, in wrapper
    return new_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3155336-404-bea4e8/openerp/addons/account_invoice_tax_required/models/account_invoice.py", line 24, in test_invoice_line_tax
    _('No Taxes Defined!\n %s') % (errors_full_string)
  UserError: (u'No Taxes Defined!\n Invoice has a line with product Dell Inspiron Laptop without OS with no taxes,Invoice has a line with product Linutop with no taxes', None)

</details>
  I'm not sure right now how can we make this...
- When making tests. We should mask the constraint on test mode for the rest of the modules. You can see an example of the inhibit here: https://github.com/OCA/bank-payment/blob/8.0/account_payment_transfer_reconcile_batch/models/payment_order.py#L28 (it contains another condition more in this case as the same method can be called from another path, but essentially it's the same), and here the way to make correctly the tests: https://github.com/OCA/bank-payment/blob/8.0/account_payment_transfer_reconcile_batch/tests/test_account_payment_transfer_reconcile_batch.py#L63

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

**Comment by angelmoya** [comment]
OK, fixed

**Review by rafaelbn — COMMENTED** [review]
I've tried to test it in runbot, but although runbot runs, it fails and I cannot open any invoice

**Comment by rafaelbn** [comment]
@angelmoya please review runbot, it's not possible to test it

File: account_invoice_tax_required/models/account_invoice.py (hunk 11 above)
Comment (pedrobaeza) [reply account_invoice_tax_required/models/account_invoice.py:15]: Ángel, this is still unattended.

**Comment by angelmoya** [comment]
@rafaelbn  I Tested on runbot, the module had not been installed, so I have installed it manually from apps menu and then I have tested that works as expect.

File: account_invoice_tax_required/models/account_invoice.py [hunk 12]
```diff
@@ -27,15 +13,15 @@ class AccountInvoice(models.Model):
...
+        for invoice_line in self.mapped('invoice_line_ids'):
+            if not invoice_line.invoice_line_tax_ids:
+                error_string = error_template % (invoice_line.name)
+                errors.append(error_string)
         if errors:
             errors_full_string = ','.join(x for x in errors)
-            raise exceptions.Warning(_('No Taxes Defined!'),
-                                     errors_full_string)
+            raise exceptions.Warning(
+                _('No Taxes Defined!\n %s') % (errors_full_string)
```
Comment (jbaudoux) [inline account_invoice_tax_required/models/account_invoice.py]: @angelmoya Can you ensure the "No Taxes Defined" translations are not broken? I'm also in favor of replacing the ',' separator by '\n' for warning readability.
`          raise exceptions.Warning(_('No Taxes Defined!') + '\n' + '\n'.join(errors))`

**Comment by pedrobaeza** [comment]
Are you going to work on this more, Ángel?

**Comment by angelmoya** [comment]
Sorry I forgot it... I can do it this week

**Comment by angelmoya** [comment]
@jbaudoux @pedrobaeza Warning message fixed to prevent broken translation, and improve with \n join

**Comment by angelmoya** [comment]
@pedrobaeza  I can not see travis log, did you know why?

**Comment by pedrobaeza** [comment]
I'm seeing it in https://travis-ci.org/OCA/account-financial-tools/jobs/215591318

**Comment by angelmoya** [comment]
OK @pedrobaeza , I don't know why Details link was not working.

Test of other module (account_move_line_purchase_info) was not working because this test create invoice without taxes, then raise user error, that is correct, but how could we prevent this?

**Comment by pedrobaeza** [comment]
See this comment: https://github.com/OCA/account-financial-tools/pull/404#discussion_r81072804

**Comment by angelmoya** [comment]
OK @pedrobaeza now test are working

**Comment by angelmoya** [comment]
@pedrobaeza can you change "Changes requested" label?

**Review by luismontalba — APPROVED** [review]

**Comment by luismontalba** [comment]
I have tested it on runbot and it works (installing the module first).

**Review by chienandalu — APPROVED** [review]
tested on runbot

File: account_invoice_tax_required/models/account_invoice.py [hunk 13]
```diff
@@ -27,20 +14,22 @@ class AccountInvoice(models.Model):
...
+                               '\n'.join(x for x in errors))
+            )
         else:
             return True
 
     @api.multi
     def invoice_validate(self):
-        self.test_invoice_line_tax()
+        if not (config['test_enable'] and
+                not self.env.context.get('test_connector')):
```
Comment (pedrobaeza) [inline account_invoice_tax_required/models/account_invoice.py]: You should rename the context variable to a more meaningful one and with no conflict with others, like `test_tax_required`

**Review by pedrobaeza — APPROVED** [review]
I have changed myself the context variable name, so merging with squash when Travis finishes

**Pull request merged**